tests/unit_tests/template.py

```diff
@@ -0,0 +1,614 @@
+# except the region-institution definition
+TEMPLATE = {
+ "simulation": {
+  "archetypes": {
+   "spec": [
+    {"lib": "agents", "name": "NullRegion"}, 
+    {"lib": "cycamore", "name": "Source"}, 
+    {"lib": "cycamore", "name": "Reactor"},
+    {"lib": "cycamore", "name": "Sink"},
+    {"lib": "d3ploy.no_inst", "name": "NOInst"}
+   ]
+  }, 
+  "control": {"duration": "15", "startmonth": "1", "startyear": "2000"}, 
+  "recipe": [
+   {
+    "basis": "mass", 
+    "name": "fresh_uox", 
+    "nuclide": [{"comp": "0.711", "id": "U235"}, {"comp": "99.289", "id": "U238"}]
+   }, 
+   {
+    "basis": "mass", 
+    "name": "spent_uox", 
+    "nuclide": [{"comp": "50", "id": "Kr85"}, {"comp": "50", "id": "Cs137"}]
+   }
+  ], 
+  "facility": [{
+   "config": {"Source": {"outcommod": "fuel",
+                         "outrecipe": "fresh_uox",
+                         "throughput": "1",
+                         "source_record_supply": "fuel"}}, 
+   "name": "source"
+  },
+  {
+   "config": {"Sink": {"in_commods": {"val":"spent_uox"},
+                       "max_inv_size": 1,
+                       "sink_record_demand": "fuel_cap"}}, 
+   "name": "sink"
+  },
+  {
+   "config": {
+        "Reactor":{
+            "assem_size":"1",
+            "cycle_time": "1", 
+            "fuel_incommods": {"val": "fuel"}, 
+            "fuel_inrecipes": {"val": "fresh_uox"}, 
+            "fuel_outcommods": {"val": "spent_uox"}, 
+            "fuel_outrecipes": {"val": "spent_uox"}, 
+            "n_assem_batch": "1", 
+            "n_assem_core": "1", 
+            "power_cap": "1", 
+            "refuel_time": "0",
+            "reactor_fuel_demand": "fuel_reactor"
+        }
+    },
+    "name": "reactor"
+   }]
+  }
+}
+
+# testA refers to if NOInst meets the increased demand by deploying more facilities.
+# testB refers to if NOInst meets the decreased demand by decommissioning existing facilities.
+
+
+# TestA Examples
+
+# Test A_1
+INIT_DEMAND = copy.deepcopy(TEMPLATE)
+INIT_DEMAND["simulation"].update({"region":{
+                         "config": {"NullRegion": "\n      "}, 
+                         "institution": {
+                          "config": {
+                           "NOInst": {
+                            "calc_method": "arma", 
+                            "demand_commod": "POWER", 
+                            "demand_std_dev": "0.0", 
+                            "growth_rate": "0.0", 
+                            "initial_demand": "1", 
+                            "prototypes": {"val": "source"}, 
+                            "steps": "1", 
+                            "supply_commod": "fuel"
+                           }
+                          }, 
+                          "name": "source_inst"
+                         }, 
+                         "name": "SingleRegion"
+                         }
+                         }
+                         )
+
+
+
+def test_a1_init_demand():
```

Comment:
As stated, this test will fail by default. It is unrealistic to assume the NO inst will know the production capabilities of a facility until that facility has operated for a time step.

Reply:
Since that means that there must always be an initial facility for the non-optimizing to work, do you think we will be able to get around this for the deterministic/stochastic optimizing methods (some sort of input in the xml)?

Reply:
I'm not sure we will ever have a clean solution around this. The institution can not make decisions to deploy unless in the tock, tick, or possibly decision phases. Therefore facilities the institution decides to deploy, can only be deployed in the second time step. It can not decided to deploy a facility before it is allowed to make a decision. Only facilities in the initial facility list will be deployed on the first time step. 

Now if this tests means will it deploy new facilities to match a demand on the second time step, the other difficulty that I mentioned before will cause issues. By default cyclus agents don't know anything about each other, therefore the institution will not know the production capabilities of one of its agents unless that agent operates. Even then it is the responsibility of the agent to inform the institution what it produced, not for the institution to ask what it produced. This is because the institution does not know what the agent is producing, or what it refers to production as (from a computational sense, the institution does not know the name of the variable on the agent that refers to its production capabilities). 

We could work around this be providing dictionary of {agent: production capability} to the institution in the input file but I don't see how this is much different than telling it to deploy facilities in the initial facilities list and determining the number deployed manually. I can make that change, but I'm not sure that will make the life of the user any easier. 

I have not thought out exactly how the DO and SO institutions will operate but I know they will not be able to make decisions about deploying facilities before the first time step (the issue mentioned in the first paragraph). If this is something that the tests need, we can find a solution. I am not convinced it will be worth the time to create it.

Reply:
I have thought about my second argument here and I think it is not correct. We may need to pass as input to the institution a set of information include agent, agent capacity, and time step agent becomes available. I'll think on this to hopefully come up with a more clever solution than that.

Reply:
This might be really far-fetched and quite labourous, but what if the 
capacity (or in general, the quantity of material / commodity an agent offers per timestep)
and the output commodity/material name were  
agent parameters (like name, lifetime)?
Wouldn't the `no_inst` able to access it then? It is a lot of work but just curious if that would be possible or not

Reply:
It would be possible in that case, but it's also limiting. D3ploy would need to scroll through all agents it owns (I don't think it would not have access to other agents) and check to see if they have that parameter. Then see if they have the right commodity to track. What about facilities that produce two materials, or three, if so how to do you indicate which rate refers to which material? Or if the capacity isn't fixed, for example it might increase over time or include some randomness. Cyclus's flexibility might be hampered if something like this were implemented.

Reply:
ah, makes sense. Thanks